server/src/main/java/org/elasticsearch/ingest/IngestService.java

```diff
@@ -408,13 +522,39 @@ private void executePipelines(
                     throw new IllegalArgumentException("pipeline with id [" + pipelineId + "] does not exist");
                 }
                 Pipeline pipeline = holder.pipeline;
+                String oldIndex = indexRequest.indices()[0];
                 innerExecute(slot, indexRequest, pipeline, onDropped, e -> {
                     if (e != null) {
                         onFailure.accept(slot, e);
                     }
 
-                    if (it.hasNext()) {
-                        executePipelines(slot, it, indexRequest, onDropped, onFailure, counter, onCompletion, originalThread);
+                    String newIndex = indexRequest.indices()[0];
+                    if (newIndex != null && oldIndex.equals(newIndex) == false && hasFinalPipeline && it.hasNext() == false) {
```

Comment:
Maybe `Objects.equal(oldIndex, newIndex) == false` instead of `newIndex != null && oldIndex.equals(newIndex) == false`

Reply:
Yep, this is good idea, I've changed it